test_runner/performance/test_bulk_insert.py

```diff
@@ -28,3 +31,38 @@ def test_bulk_insert(neon_with_baseline: PgCompare):
 
             env.report_peak_memory_use()
             env.report_size()
+
+    # When testing neon, also check how long it takes the pageserver to reingest the
+    # wal from safekeepers. If this number is close to total runtime, then the pageserver
+    # is the bottleneck.
+    if isinstance(env, NeonCompare):
+        measure_recovery_time(env, test_output_dir)
+
+
+def measure_recovery_time(env: NeonCompare, test_output_dir):
+    client = env.env.pageserver.http_client()
+    pg_version = PgVersion(client.timeline_detail(env.tenant, env.timeline)["pg_version"])
+
+    def get_lsn():
+        return client.timeline_detail(env.tenant, env.timeline)["last_record_lsn"]
+
+    target_lsn = get_lsn()
+
+    # Stop pageserver and remove tenant data
+    env.env.pageserver.stop()
+    timeline_dir = (
+        test_output_dir / "repo" / "tenants" / str(env.tenant) / "timelines" / str(env.timeline)
+    )
+    shutil.rmtree(timeline_dir)
+
+    # Start pageserver
+    env.env.pageserver.start()
+
+    # Measure recovery time
+    with env.record_duration("wal_recovery"):
+        # Create the tenant, which will start walingest
+        client.timeline_create(pg_version, env.tenant, env.timeline)
+
+        # wait for lsn
+        while get_lsn() < target_lsn:
+            time.sleep(0.1)
```

Comment:
If you'd branch from env.timeline on this lsn, it would wait for the lsn to catch up.

Reply:
Could also add a `wait_lsn` http endpoint, which would be useful if we ever need to (again) do this in prod. If it returned chunked text/plain response, we could have a progress update every second, no matter how long the whole operation took.

Reply:
Well, perhaps the 0.1s is fine.